compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -158,13 +158,35 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
     #[instrument(level = "debug", skip(self), ret)]
     fn compute_alias_eq_goal(
         &mut self,
-        goal: Goal<'tcx, (ty::Term<'tcx>, ty::Term<'tcx>)>,
+        goal: Goal<'tcx, (ty::Term<'tcx>, ty::Term<'tcx>, ty::AliasRelationDirection)>,
     ) -> QueryResult<'tcx> {
         let tcx = self.tcx();
 
-        let evaluate_normalizes_to = |ecx: &mut EvalCtxt<'_, 'tcx>, alias, other| {
+        let evaluate_normalizes_to = |ecx: &mut EvalCtxt<'_, 'tcx>, alias, other, direction| {
             debug!("evaluate_normalizes_to(alias={:?}, other={:?})", alias, other);
-            let r = ecx.probe(|ecx| {
+            let result = ecx.probe(|ecx| {
+                let other = match direction {
+                    // This is purely an optimization.
+                    ty::AliasRelationDirection::Equate => other,
+
+                    ty::AliasRelationDirection::Subtype | ty::AliasRelationDirection::Supertype => {
+                        let fresh = ecx.next_term_infer_of_kind(other);
+                        let (sub, sup) = if direction == ty::AliasRelationDirection::Subtype {
+                            (fresh, other)
+                        } else {
+                            (other, fresh)
+                        };
+                        ecx.add_goals(
+                            ecx.infcx
+                                .at(&ObligationCause::dummy(), goal.param_env)
+                                .sub(DefineOpaqueTypes::No, sub, sup)?
+                                .into_obligations()
+                                .into_iter()
+                                .map(|o| o.into()),
```

Comment:
that's the second place that uses `sub` in the solver :thinking: maybe we should just add it to `EvalCtxt` :sweat_smile: feeling pretty confident that we're able to catch any misuses of it.

I generally feel like the auto ping works fairly well and expect the solver to not change too much over time once it is stable.